scipy/sparse/base.py

```diff
@@ -125,8 +116,8 @@ def reshape(self, shape, order='C'):
         --------
         np.matrix.reshape : NumPy's implementation of 'reshape' for matrices
         """
-        raise NotImplementedError("Reshaping not implemented for %s." %
-                                  self.__class__.__name__)
+        return (self.tocoo(copy=copy).reshape(*args, order=order, copy=True)
+                .asformat(self.format))
```

Comment:
Some existing methods have a `format=None` keyword argument, where `None` means to return the most efficient / convenient format. Might be useful here.

Reply:
I'm not opposed to this idea, but I couldn't find an example of the format option in `base.py`. Maybe we can defer this to a larger refactoring for including this option.

I did find that converting back to the same format is not common, however. The default for most methods is `return self.tocsr().some_operation()`. I think I'll try to write a `reshape` method for CSR and use that as the default instead. If users got back a different type from `reshape`, I think they would be OK with a CSR, but probably not a COO.

Reply:
The more I looked at it, the COO matrix is a first class citizen of the sparse package and a perfectly fine object to return. I looked at the CSC reshape implementation in the Octave source, but couldn't figure out how to implement that in Python that would likely be faster than just using the existing C code to cast to COO and back. If someone knows how to do C work, they could probably copy the algorithm from Octave and wire it into `_sparsetools`.

Reply:
Please don't look at the Octave source if you intend to contribute similar things here.  Octave is GPL and Scipy is not compatible with that license.

Reply:
That is a good point. I guess I'll leave the clean room implementation of CSR reshape to someone else.